controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +124,296 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
 	return result, nil
 }
 
-func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Consumer Mode")
+func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Result, error) {
+	r.log.Info("Running StorageClassClaim controller in Consumer Mode")
+
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		r.storageCluster.Spec.ExternalStorage.StorageProviderEndpoint,
+		10*time.Second,
+	)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Close client-side connections.
+	defer providerClient.Close()
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+
+		// Check if finalizers are present, if not, add them.
+		if !contains(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer) {
+			storageClassClaimRef := klog.KRef(r.storageClassClaim.Name, r.storageClassClaim.Namespace)
+			r.log.Info("Finalizer not found for StorageClassClaim. Adding finalizer.", "StorageClassClaim", storageClassClaimRef)
+			r.storageClassClaim.SetFinalizers(append(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer))
+			if err := r.update(r.storageClassClaim); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to update StorageClassClaim [%v] with finalizer: %s", storageClassClaimRef, err)
+			}
+		}
+
+		// Creation phase.
+		r.storageClassClaim.Status.Phase = ocsv1alpha1.StorageClassClaimCreating
+
+		// storageClassClaimStorageType is the storage type of the StorageClassClaim
+		var storageClassClaimStorageType providerclient.StorageType
+		switch r.storageClassClaim.Spec.Type {
+		case "blockpool":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestBlockpool
+		case "sharedfilesystem":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestSharedfilesystem
+		}
+
+		// Call the `FulfillStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+		_, err = providerClient.FulfillStorageClassClaim(
+			r.ctx,
+			r.storageCluster.Status.ExternalStorage.ConsumerID,
+			r.storageClassClaim.Name,
+			r.storageClassClaim.Spec.EncryptionMethod,
+			storageClassClaimStorageType,
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to initiate fulfillment of StorageClassClaim: %v", err)
+		}
```

Comment:
fullfill should get called only if the status.Phase is empty?

Reply:
i think this should be fine as it helps with update functionality which we don't have currently.

Reply:
@Madhu-1 this is wrong, phase should not dictate skipping of steps. 
Because every reconcile should go through all steps. 

Mainly because there might be other steps that need to be done on every reconcile the preceded the fulfilment - for example a validation phase (which is missing from this PR)

If you need an indication that fulfillment already happened for this resource then we need to keep that information as a separate field on the status (something like, is fulfilled). 

I have asked @rexagod to postpone this additional CR status change for a future PR